apps/heft/src/plugins/TypeScriptPlugin/TypeScriptBuilder.ts

```diff
@@ -141,6 +158,57 @@ export class TypeScriptBuilder extends SubprocessRunnerBase<ITypeScriptBuilderCo
       this._configuration.terminalPrefixLabel
     );
 
+    // Determine the compiler version
+    const compilerPackageJsonFilename: string = path.join(
+      this._configuration.typeScriptToolPath,
+      'package.json'
+    );
+    const packageJson: IPackageJson = JsonFile.load(compilerPackageJsonFilename);
+    this._typescriptVersion = packageJson.version;
+    const parsedVersion: semver.SemVer | null = semver.parse(this._typescriptVersion);
+    if (!parsedVersion) {
+      throw new Error(
+        'Unable to parse version "${this._typescriptVersion}" for TypeScript compiler package in: ' +
+          compilerPackageJsonFilename
+      );
+    }
+    this._typescriptParsedVersion = parsedVersion;
+
+    // Detect what features this compiler supports.  Note that manually comparing major/minor numbers
+    // loosens the matching to accept prereleases such as "3.6.0-dev.20190530"
+    this._capabilities = {
+      incrementalProgram: false
+    };
+    if (
+      this._typescriptParsedVersion.major > 3 ||
+      (this._typescriptParsedVersion.major === 3 && this._typescriptParsedVersion.minor >= 6)
+    ) {
+      this._capabilities.incrementalProgram = true;
+    }
+
+    // TypeScript 2.9 introduced "ts.getConfigFileParsingDiagnostics()" which we currently rely on.
+    // More fixups are required to support older versions.  We won't do that work unless someone requests it.
+    if (
+      this._typescriptParsedVersion.major < 2 ||
+      (this._typescriptParsedVersion.major === 2 && this._typescriptParsedVersion.minor < 9)
+    ) {
+      // We don't use TerminalProviderSeverity.warning here because, if the person wants to take their chances with
+      // a seemingly unsupported compiler, their build should be allowed to succeed.
+      this._terminalProvider.write(
+        `The TypeScript compiler version ${this._typescriptVersion} is very old` +
+          ` and has not been tested with Heft; it may not work correctly.` +
+          this._terminalProvider.eolCharacter,
+        TerminalProviderSeverity.log
+      );
+    } else if (this._typescriptParsedVersion.major > 3) {
+      this._terminalProvider.write(
+        `The TypeScript compiler version ${this._typescriptVersion} is newer` +
+          ` than the latest version that was tested with Heft; it may not work correctly.` +
+          this._terminalProvider.eolCharacter,
+        TerminalProviderSeverity.log
+      );
```

Comment:
```suggestion
      this.._typescriptTerminal.writeWarningLine(
        `The TypeScript compiler version ${this._typescriptVersion} is very old` +
          ` and has not been tested with Heft; it may not work correctly.`
      );
    } else if (this._typescriptParsedVersion.major > 3) {
      this._typescriptTerminal.writeWarningLine(
        `The TypeScript compiler version ${this._typescriptVersion} is newer` +
          ` than the latest version that was tested with Heft; it may not work correctly.`
      );
```

Reply:
Fixed